File: DSFM-Java/DSFM.java
/*
  Java 17
  Compiler Used: Amazon Correto JDK 17.0
  IDE Used: VS Code

  Remade C++ DSFM Sim with Java
*/
import java.io.*;
import java.util.Arrays;
import java.util.Scanner;

public class DSFM {
    public static String removeSeparators(String x) {
        return x.replace(",", "").replace(" ", "").replace("(", "").replace(")", "");
    }

    public static int findIndex(String a, char c) {
        int index = a.indexOf(c);
        return index != -1 ? index : -1;
    }

    public static void main(String[] args) {
        try {
            Scanner sim = new Scanner(new File("dsfm.txt"));

            // Grab Alphabet language
            String alphabet = removeSeparators(sim.nextLine());
            int numL = alphabet.length();
            char[] L = alphabet.toCharArray();
            int numLang = L.length;

            // Get num states
            int numStates = Integer.parseInt(sim.nextLine().trim());

            // Get accepting states
            String strAccept = removeSeparators(sim.nextLine());
            boolean[] A = new boolean[10];
            Arrays.fill(A, false);
            for (int i = 0; i < strAccept.length(); i++) {
                char k = strAccept.charAt(i);
                int indexAccept = strAccept.indexOf(k);
                if (indexAccept != -1) {
                    A[i] = true;
                }
            }

            // Transitions
            char[][] goTo = new char[numStates][numL];
            int state = 0;
            int lang = 0;
            for (int i = 0; i < numStates; i++) {
                int point = 2;
                String temp = removeSeparators(sim.nextLine());
                for (int j = 0; j < numLang; j++) {
                    char nextState = temp.charAt(point);
                    goTo[state][lang] = nextState;
                    point = point + 3;
                    lang++;
                }
                lang = 0;
                state++;
            }
            sim.close();

            // User interaction
            Scanner inputScanner = new Scanner(System.in);
            char retry;
            do {
                String input;
                int currentState = 0;
                boolean valid = true;
                boolean accepted = false;
                int index;
                System.out.println("Enter a string using alphabet");
                input = inputScanner.nextLine();
                int sizeInput = input.length();
                for (int i = 0; i < sizeInput; i++) {
                    char test = input.charAt(i);
                    index = findIndex(alphabet, test);
                    if (index != -1) {
                        currentState = goTo[currentState][index] - '0';
                        System.out.println("\nStep: " + currentState);
                    } else {
                        System.out.println("\nError certain characters not in Language");
                        valid = false;
                        break;
                    }
                }
                accepted = A[currentState];

                if (accepted && valid) {
                    System.out.println("\nYour string was accepted!");
                } else {
                    System.out.println("\nYour string was NOT accepted!");
                }
                // Prompt user to repeat program
                System.out.println("Do you want to try another string? (Y/N)");
                retry = inputScanner.nextLine().charAt(0);
                System.out.println("\n\n\n");
            } while (retry == 'y' || retry == 'Y');
        } catch (FileNotFoundException e) {
            System.out.println("Failed to retrieve file. Try again (Make sure file is \"testRun.txt\")\nPress ENTER to exit...");
            new Scanner(System.in).nextLine();
        }
    }
}

File: DSFM-C++(Depricated)/DSFM.cpp
/*
  Course: CS 41000
  Project: DSFM Project
  Name:Romulo Gutierrez
  Email:gutie136@pnw.edu
  
  C++
  Compiler Used: MinGW GCC
  IDE Used: Programmer's Notepad + OnlineGDB
*/

// IMPORTANT NOTE:
// This program was compiled using gcc compiler with Programmer's Notepad on Windows 10.
// Works properly when using OnlineGDB with testRun.txt file
// (Depricated) Newer Java version working properly

#include <iostream>
#include <cstring>
#include <string>
#include <cstdlib>
#include <fstream>
#include <algorithm>

using namespace std;

string removeSeperators(string x)
{ // remove the seperators
	x.erase(remove(x.begin(), x.end(), ','), x.end());
	x.erase(remove(x.begin(), x.end(), ' '), x.end());
	x.erase(remove(x.begin(), x.end(), '('), x.end());
	x.erase(remove(x.begin(), x.end(), ')'), x.end());
	return x;
}
int findIndex(string input, char c)
{ // finds index of strings
	int index = input.find(c);
	if (index != string::npos)
	{return index;}
	else {return -1;}
}

int main()
{
	ifstream file("testRun.txt");
	if (file.fail())
	{
		cout << "Failed to retrieve file. Try again (Make sure file is \"testRun.txt\")\nPress any key to exit..." << endl;
		cin.get();											 
		exit(0);
	}
	// Grab Alphabet language
	string alphabet;
	getline(file, alphabet);
	alphabet = removeSeperators(alphabet);
	int alphabetSize = alphabet.size();
	char alphabetArray[alphabetSize];
	strcpy(alphabetArray, alphabet.c_str());
	int numLang = sizeof(alphabetArray) / sizeof(alphabetArray[0]);
	
	// Get num states
	string numStatesChar;
	getline(file, numStatesChar);
	int numStates = atoi(numStatesChar.c_str());

	// Get accepting states -- Good
	string strStateAccept;
	getline(file, strStateAccept);
	strStateAccept = removeSeperators(strStateAccept);
	int sizeAccept = strStateAccept.size();
	bool acceptStates[10];
	fill_n(acceptStates, 10, false);
	for (int i = 0; i < sizeAccept; i++)
	{
		char k = strStateAccept.at(i);
		int indexAccept = strStateAccept.find(k);
		if (static_cast<int>(string::npos) != indexAccept)
		{
			acceptStates[i] = true;
		}
	}

	// Transitions
	char transitionMatrix[numStates][alphabetSize];
	string strTransition;
	int currentState = 0;
	int currentLang = 0;
	char nextState;
	int point = 2;
	for (int i = 0; i < numStates; i++)
	{
		point = 2;
		getline(file, strTransition);
		strTransition = removeSeperators(strTransition);
		// cout <<  "Transition: " << strTransition << endl;
		for (int j = 0; j < numLang; j++)
		{
			nextState = strTransition.at(point);
			transitionMatrix[currentState][currentLang] = nextState;
			// cout << "Transit to: "<< transitionMatrix[currentState][currentLang] << endl;
			point = point + 3;
			currentLang++;
		}
		currentLang = 0;
		currentState++;
		strTransition.erase();
	}
	file.close();

	// User interaction
	char retry;
	do
	{
		string userInput;
		int sizeInput;
		int currentState = 0;
		bool isValid = true;
		bool isAccepted = false;
		int index;
		cout << "Enter a string using alphabet" << endl;
		cin >> userInput;
		int userInputSize = userInput.size();
		sizeInput = sizeof(userInput) / sizeof(userInput[0]);
		for (int i = 0; i < userInputSize; i++)
		{
			char currentChar = userInput.at(i);
			index = findIndex(alphabet, currentChar);
			if (index != -1)
			{
				currentState = transitionMatrix[currentState][index] - '0';
				cout << endl
					 << "Step: " << currentState;
			}
			else
			{
				cout << endl
					 << "Error certain characters not in Language";
				isValid = false;
				break;
			}
		}
		isAccepted = acceptStates[currentState];
	
		if (isAccepted == true && isValid == true)
		{
			cout << endl
				 << "Your string was accepted!" << endl;
		}
		else
		{
			cout << endl
				 << "Your string was NOT accepted!" << endl;
		}
		// Prompt user to repeat program
		cout << "Do you want to try another string? (Y/N)" << endl;
		cin >> retry;
		cout << endl
			 << endl
			 << endl;
	} while (retry == 'y' || retry == 'Y');
	return 0;
	}